[SPARK-4229] Create hadoop configuration in a consistent way

koeninger opened the pull request.

koeninger committed:
SPARK-4229 use SparkHadoopUtil.get.conf so that hadoop properties are copied from spark config
Resolved conflicts in favor of master.

Conflicts:
	streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala
	streaming/src/main/scala/org/apache/spark/streaming/dstream/PairDStreamFunctions.scala

koeninger committed:
SPARK-4229 document handling of spark.hadoop.* properties

koeninger committed:
SPARK-4229 fix broken table in documentation, make hadoop doc formatting match that of runtime env

koeninger committed:
Merge branch 'master' of https://github.com/apache/spark into SPARK-4229-master

koeninger committed:
[SPARK-4229] per tdas and JoshRosen, use sparkContext.hadoopConfiguration where possible, synchronize new Configuration

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaPairDStream.scala:
```diff
@@ -789,7 +790,7 @@ class JavaPairDStream[K, V](val dstream: DStream[(K, V)])(
       keyClass: Class[_],
       valueClass: Class[_],
       outputFormatClass: Class[_ <: NewOutputFormat[_, _]],
-      conf: Configuration = new Configuration) {
```
The scope of this PR is pretty wide in terms of the number of classes it touches, causing issues as different places needs to be handled differently. If you considered moving this sort of changes (`new Configuration` to `sparkContext.hadoopConfiguration`) into a different PR that might be easier to get in.

koeninger replied:
Based on what Marcelo Vanzin said on the dev list when I brought this issue
up, the only reason the problem was still around for me to run into is
because he changed some of the uses of new Configuration but not all of
them.

I agree it's used in a lot of different places, but I'm not sure how
piecemeal fixes to only some of the places is helpful to users. Were there
still specific concerns about particular classes?

On Sun, Jan 4, 2015 at 6:28 AM, Tathagata Das <notifications@github.com>
wrote:

> In
> streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaPairDStream.scala
> <https://github.com/apache/spark/pull/3543#discussion-diff-22438855>:
>
> > @@ -789,7 +790,7 @@ class JavaPairDStream[K, V](val dstream: DStream[(K, V)])(
> >        keyClass: Class[_],
> >        valueClass: Class[_],
> >        outputFormatClass: Class[_ <: NewOutputFormat[_, _]],
> > -      conf: Configuration = new Configuration) {
>
> The scope of this PR is pretty wide in terms of the number of classes it
> touches, causing issues as different places needs to be handled
> differently. If you considered moving this sort of changes (new
> Configuration to sparkContext.hadoopConfiguration) into a different PR
> that might be easier to get in.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/3543/files#r22438855>.
>

vanzin commented:
Just for posterity: I think a couple of changes sneaked in between my original change being sent as a PR and it being commited, making my code miss some `Configuration` instantiations.

At that time, I explicitly avoided changing default arguments to a few methods (since my thinking was that since it's an argument, the user should know what he's doing). But I don't really have an opinion about what's the right approach there, and changing it is fine with me.

I also don't have enough background to comment on the thread-safety issues (others have looked at it in much more depth than I have)...

marmbrus commented:
ping.  Whats the status here?

koeninger commented:
I think its mostly a question of whether committers are comfortable with a
PR that changes all of the uses of new Configuration.

At this point it'd probably need another audit of the code to see if there
are more uses, but that's mostly mechanical.

On Tue, Mar 17, 2015 at 10:41 PM, Michael Armbrust <notifications@github.com
> wrote:

> ping. Whats the status here?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/3543#issuecomment-82725434>.
>

AmplabJenkins commented:
Can one of the admins verify this patch?

andrewor14 commented:
Another ping. @koeninger @tdas @JoshRosen should we move forward with this patch, or close it since it's mostly gone stale at this point?

koeninger commented:
As far as I know, its still an issue - by default, any checkpoint that
relies on hdfs config (e.g. s3 password) won't recover
On Jun 18, 2015 6:55 PM, "andrewor14" <notifications@github.com> wrote:

> Another ping. @koeninger <https://github.com/koeninger> @tdas
> <https://github.com/tdas> @JoshRosen <https://github.com/JoshRosen>
> should we move forward with this patch, or close it since it's mostly gone
> stale at this point?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/3543#issuecomment-113321519>.
>

tdas commented:
Hey @koeninger, looking at this patch again, I would like to absorb the streaming changes at the very least. Those issues still exist in streaming, and would be a good fix to have. So mind closing this PR and issuing a new PR with only the fixes to the streaming API?

tdas commented:
@koeninger Ping.

koeninger commented:
Just to be clear, are we talking about removing just the one-line changes to SQLContext and JavaSQLContext?

Everything else in the PR I think is necessary in order to make the changes in streaming.

tdas commented:
Otherway round. Just keep the changes in StreamingContext, DStream, and PairDStreamFunctions.

koeninger commented:
Except that those streaming changes call into SparkHadoopUtil, which was changed in that PR for thread safety reasons.  HadoopRDD was changed so there was only 1 lock being used.  At that point the only thing left is doc changes and the sql changes.

tdas commented:
Aah right, makes sense. That definitely complicates things, because that is the hard questions, whether to put that lock or not. @JoshRosen is the best person to answer that. Unfortunately he is swamped :(

JoshRosen commented:
Haven't dug in to this in detail yet, but it's possible that the bug that motivated the `CONFIGURATION_INSTANTIATION_LOCK` is no longer relevant to us because we no longer support the affected Hadoop versions.  It would be great if someone more familiar with Hadoop version numbering / JIRA conventions could look at the Hadoop JIRA ticket to figure this out.  If it turns out that it only affects pre-Hadoop 1.2.1 versions, then we might be able to just remove that lock entirely.

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #38584 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38584/consoleFull) for   PR 3543 at commit [`bfc550e`](https://github.com/apache/spark/commit/bfc550ef0b7b535adb0aa019f30dd4771c24aece).
 * This patch **does not merge cleanly**.

SparkQA commented:
[Test build #38584 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38584/consoleFull) for   PR 3543 at commit [`bfc550e`](https://github.com/apache/spark/commit/bfc550ef0b7b535adb0aa019f30dd4771c24aece).
 * This patch **fails to build**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

koeninger commented:
If we're talking about this issue https://issues.apache.org/jira/browse/HADOOP-11209 unless there's something arcade about hadoop's jira, it looks like that was only resolved in April for 2.7

@tdas if you think we're better off / not worse off with at least having the streaming-only changes in for spark 1.5, I can put in a narrower PR for that and we can punt on the thread safety issues for now

tdas commented:
Yeah, may be that is a good idea for now. In fact, SparkHadoopUtil.get.conf calls newConfiguration only once, and then the configuration is used everywhere. So the newConfiguration() will be called only every once in the lifetime of the application, and the likelihood of the race condition causing a problem here is really small. So I think its fine for now to just address this. 

The way I would do this is to make the JIRA specific to streaming only (set component and title accordingly). And file a separate JIRA (if not already present) for a possible problem in newConfiguration() linking it to the Hadoop JIRA. Does that make sense? @JoshRosen Any thoughts?

koeninger commented:
Changing this jira to be streaming only and making another for thread safety issues still leaves all the inconsistent calls to new Configuration in SQL, and probably other places (at a quick grep, external/flume, external/twitter, and maybe core).

Ill get a PR with changes only to streaming/, let me know what you guys want to do as far as jiras

koeninger commented:
streaming only pr is at https://github.com/apache/spark/pull/7772

tdas commented:
Fair point. How about make subtasks of the JIRA for different components, and then use those JIRA ids?

koeninger commented:
Added subtasks, changed the title of https://github.com/apache/spark/pull/7772 to refer to the streaming subtask jira ID.  Let me know if you see anything on that that needs tweaking before the 1.5 freeze date

tdas commented:
Okay #7772 has been merged. Mind removing the streaming changes from this PR to make this cleaner?

andrewor14 commented:
@koeninger would you mind updating this patch per @tdas' suggestion?

koeninger commented:
@andrewor14 master has diverged sufficiently from this PR that I don't think it's useful to keep it merge-able.  If we think someone's willing to accept the changes to core and sql those subtasks should be revisited with this general approach as a basis.

andrewor14 commented:
OK, makes sense. Can you close this PR for now then? If there's interest we can always reopen it against the latest master.
